src/ducks/select/index.js

```diff
@@ -69,18 +71,34 @@ const reducer = (state = INITIAL_STATE, action) => {
 };
 
 function applyToggleItem(state, action) {
-  let { stateKey, id } = action.payload;
-  let list = state[stateKey] || [];
-  let index = list.indexOf(id);
-  let selectedItems;
-
-  if (index !== -1) {
-    selectedItems = removeItem(list, index);
-  } else {
-    selectedItems = addItem(list, id);
+  const { stateKey, id, event, allIds } = action.payload;
+  const currentSelection = state[stateKey] && state[stateKey].selectedItems ?
+    state[stateKey].selectedItems :
+    [];
+  const index = currentSelection.indexOf(id);
+  const isAlreadySelected = index !== -1;
+  if (!isAlreadySelected && allIds && allIds.length && event && event.shiftKey) {
+    const lastSelectedItem = state[stateKey].lastSelectedItem || id;
+    const selectedRange = getSelectedRange(allIds, id, lastSelectedItem);
+    const selectedItems = uniq([...currentSelection, ...selectedRange]);
+    return { ...state, [stateKey]: { selectedItems, lastSelectedItem } };
   }
+  const selectedItems = isAlreadySelected
+    ? removeItem(currentSelection, index)
+    : addItem(currentSelection, id);
+  const lastSelectedItem = isAlreadySelected
+    ? state[stateKey].lastSelectedItem
+    : id;
+  return { ...state, [stateKey]: { selectedItems, lastSelectedItem } };
+}
 
-  return { ...state, [stateKey]: selectedItems };
+function getSelectedRange(allIds, id, lastSelectedItem) {
+  const lastSelectedItemIndex = allIds.indexOf(lastSelectedItem);
+  const currentSelectedItemIndex = allIds.indexOf(id);
+  const firstIndex = Math.min(lastSelectedItemIndex, currentSelectedItemIndex);
+  const lastIndex = Math.max(lastSelectedItemIndex, currentSelectedItemIndex);
+  const selectedRange = allIds.slice(firstIndex, lastIndex + 1);
+  return selectedRange;
```

Comment:
it's nice to name things (the selectedRange variable), but here I'd assume that the reader knows which method he/she is reading, and the method name is quite revealing about what is returned ;)

So I'd recommend just doing: `return allIds.slice(firstIndex, lastIndex + 1);`

Reply:
Oops, yeah, that was a refactoring leftover, I'm afraid.